### PR TITLE
[batch] Give users in billing project access to batches, jobs

### PR DESCRIPTION
**Pull request by jigold**
There might be an opportunity to also push NOT DELETED into the decorator as well. I chose the decorator route because it's cleaner and I expect a more complicated query for checking permissions in the future.

### CODE REVIEW

**Comment by jigold** [comment]
I can't figure out why I'm getting an error in one test. But I also am not sure what to do with the `/batches` endpoint. I want the UI default to only show you your batches with the default query string 'user:jigold`. However, what should the REST endpoint be? All batches in all billing projects you have access for?

File: batch/batch/front_end/front_end.py [hunk 1]
```diff
@@ -115,6 +115,49 @@ async def wrapped(request, userdata, *args, **kwargs):
...
+    return record is not None
+
+
+def rest_billing_project_users_only(fun):
+    @rest_authenticated_users_only
+    @wraps(fun)
+    async def wrapped(request, userdata, *args, **kwargs):
+        permitted_user = await _batch_billing_project_user(request, userdata)
+        if not permitted_user:
+            raise web.HTTPUnauthorized()
```
Comment (danking) [inline batch/batch/front_end/front_end.py:143]: 401 Unauthorized doesn't seem right. 401 [suggests that the client has not sent the proper authentication information](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/401).

The question is then should we return 403 or 404. I think our policy is to always return 404.

File: batch/batch/front_end/front_end.py [hunk 2]
```diff
@@ -250,17 +293,16 @@ async def _query_batch_jobs(request, batch_id):
 
 @routes.get('/api/v1alpha/batches/{batch_id}/jobs')
 @prom_async_time(REQUEST_TIME_GET_JOBS)
-@rest_authenticated_users_only
-async def get_jobs(request, userdata):
+@rest_billing_project_users_only
+async def get_jobs(request, userdata):  # pylint: disable=unused-argument
     batch_id = int(request.match_info['batch_id'])
```
Comment (danking) [inline batch/batch/front_end/front_end.py:298]: The fas that `rest_billing_project_users_only` does not pass the batch_id to get_jobs means that a refactoring can easily break our protection mechanisms. Let's make both `_billing_project_users_only` pass the approved batch_id as an argument to the wrapped method.

File: batch/batch/front_end/front_end.py [hunk 3]
```diff
@@ -534,10 +575,9 @@ async def _query_batches(request, user):
 
 @routes.get('/api/v1alpha/batches')
 @prom_async_time(REQUEST_TIME_GET_BATCHES)
-@rest_authenticated_users_only
-async def get_batches(request, userdata):
-    user = userdata['username']
-    batches, last_batch_id = await _query_batches(request, user)
+@rest_billing_project_users_only
```
Comment (danking) [inline batch/batch/front_end/front_end.py:578]: This request doesn't have a batch_id, right?

File: batch/batch/front_end/front_end.py [hunk 4]
```diff
@@ -427,20 +469,19 @@ async def _get_full_job_status(app, record):
...
     return web.json_response(job_log)
 
 
-async def _query_batches(request, user):
+async def _query_batches(request):
     db = request.app['db']
 
-    where_conditions = ['user = %s', 'NOT deleted']
-    where_args = [user]
+    where_conditions = ['NOT deleted']
```
Comment (danking) [inline batch/batch/front_end/front_end.py:483]: This method appears to now have no restrictions at all?

File: batch/batch/front_end/front_end.py [hunk 5]
```diff
@@ -1112,10 +1144,9 @@ async def ui_delete_batch(request, userdata):
 
 @routes.get('/batches', name='batches')
 @prom_async_time(REQUEST_TIME_GET_BATCHES_UI)
-@web_authenticated_users_only()
+@web_billing_project_users_only()
 async def ui_batches(request, userdata):
-    user = userdata['username']
-    batches, last_batch_id = await _query_batches(request, user)
+    batches, last_batch_id = await _query_batches(request)
```
Comment (danking) [inline batch/batch/front_end/front_end.py:1149]: this now appears to return all batches for all billing projects

**Review by danking — CHANGES_REQUESTED** [review]

File: batch/batch/front_end/front_end.py (hunk 2 above)
Comment (jigold) [reply batch/batch/front_end/front_end.py:298]: Can you give an example?

File: batch/batch/front_end/front_end.py (hunk 3 above)
Comment (jigold) [reply batch/batch/front_end/front_end.py:578]: No. It's broken. See the comment I made above about making the default query string `f'user:{user}'`

File: batch/batch/front_end/front_end.py (hunk 4 above)
Comment (jigold) [reply batch/batch/front_end/front_end.py:483]: Again, see comment above. We have to do something different for query_batches or queries that don't have a specific batch idea like query_billing (which I already implemented previously).

File: batch/batch/front_end/front_end.py (hunk 5 above)
Comment (jigold) [reply batch/batch/front_end/front_end.py:1149]: See comment above.

File: batch/test/test_accounts.py [hunk 6]
```diff
@@ -398,3 +398,101 @@ async def test_billing_limit_tiny(make_client, dev_client, new_billing_project):
...
+
+    try:
+        await dev_client._patch(f'/api/v1alpha/batches/{b.id}/cancel')
+    except aiohttp.ClientResponseError as e:
+        assert e.status == 401, e
+    else:
+        assert False, 'expected error'
+
+    try:
+        await dev_client._delete(f'/api/v1alpha/batches/{b.id}/delete')
```
[the remark this answers is not in the document]
Comment (jigold) [reply batch/test/test_accounts.py:492]: That doesn't make sense. It should error out on line 443. I just added delete and cancel as methods on batch_client. At some point we should rewrite the whole thing.

File: batch/batch/front_end/front_end.py (hunk 2 above)
Comment (danking) [reply batch/batch/front_end/front_end.py:298]: Like this:
```
@rest_billing_project_users_only
 async def get_jobs(request, userdata, batch_id):  # pylint: disable=unused-argument
```
```
async def user_can_access(db: Database, batch_id: int, username: str):
     return None is not await db.select_and_fetchone(
         '''
 SELECT id
 FROM batches
 LEFT JOIN billing_project_users ON batches.billing_project = billing_project_users.billing_project
 WHERE id = %s AND billing_project_users.`user` = %s;
 ''',
         (batch_id, username))

def rest_billing_project_users_only(fun):
     @rest_authenticated_users_only
     @wraps(fun)
     async def wrapped(request, userdata, *args, **kwargs):
         batch_id = request['batch_id']
         if not await user_can_access(request['db'], batch_id, userdata['username']):
             raise web.HTTPNotFound()
        return await fun(request, userdata, batch_id, *args, **kwargs)
     return wrapped
```

File: batch/test/test_accounts.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (danking) [reply batch/test/test_accounts.py:492]: ah, right, so we can just construct a batch object then, right?
```
Batch(dev_client, b.id, b.attributes, b.n_jobs)
```
I just mean that we should use the API we've built. That forces us to make sure the API works well.

File: batch/test/test_accounts.py [hunk 7]
```diff
@@ -398,3 +398,101 @@ async def test_billing_limit_tiny(make_client, dev_client, new_billing_project):
     batch = await batch.submit()
     batch = await batch.wait()
     assert batch['state'] == 'cancelled', batch
+
+
+async def test_all_billing_project_users_operations(make_client, dev_client, new_billing_project):
```
Comment (danking) [inline batch/test/test_accounts.py:403]: and this is something like `test_user_can_access_batch_made_by_other_user_in_shared_billing_project`, yeah?

File: batch/batch/front_end/front_end.py [hunk 8]
```diff
@@ -471,6 +509,18 @@ async def _query_batches(request, user):
 ((batches.id) IN
  (SELECT batch_id FROM batch_attributes
   WHERE `key` = %s))
+'''
+            args = [k]
+        elif t.startswith('user:'):
```
Comment (danking) [inline batch/batch/front_end/front_end.py:514]: Can you add these new options to the documentation in `batches.html`?

File: batch/batch/front_end/front_end.py [hunk 9]
```diff
@@ -448,7 +486,7 @@ async def _query_batches(request, user):
         where_conditions.append('(id < %s)')
         where_args.append(last_batch_id)
 
-    q = request.query.get('q', '')
+    q = request.query.get('q', f'user:{user}')
```
Comment (danking) [inline batch/batch/front_end/front_end.py:489]: This needs to be consistent with the UI. Currently the UI will show no search term. We should probably explicitly update the `request.query` term in both `get_batches` and `ui_batches`. That ensures that when we regenerate the page in `ui_batches`, we'll use the right query term.

**Review by danking — DISMISSED** [review]

File: ci/Makefile [hunk 10]
```diff
@@ -13,7 +13,7 @@ PYTHON := PYTHONPATH=$${PYTHONPATH:+$${PYTHONPATH}:}$(EXTRA_PYTHONPATH) python3
 check:
 	$(PYTHON) -m flake8 ci
 	$(PYTHON) -m pylint --rcfile ../pylintrc ci --score=n
-	$(PYTHON) -m black ci --check --line-length=120 --skip-string-normalization
+	$(PYTHON) -m black ci --check --diff --line-length=120 --skip-string-normalization
```
Comment (danking) [inline ci/Makefile:16]: what's this?

**Review by danking — DISMISSED** [review]

File: ci/Makefile (hunk 10 above)
Comment (jigold) [reply ci/Makefile:16]: I was getting errors in this check. `--diff` actually gives you the difference on what your original files looked like and what the rewrite looks like. I think this is fine to keep in so someone else who modifies CI doesn't have to look for this option if this check fails.

File: batch/batch/front_end/templates/batches.html [hunk 11]
```diff
@@ -25,6 +25,8 @@ <h1>Batches</h1>
 	  <ul>
 	    <li>k=v - batches with an attribute with key k and value v</li>
 	    <li>has:k - batches that have an attribute with key k</li>
```
Comment (daniel-goldstein) [inline batch/batch/front_end/templates/batches.html:27]: This line confuses me now. Should this be `col:v` and not be attribute?

File: batch/batch/front_end/templates/batches.html [hunk 12]
```diff
@@ -25,6 +25,8 @@ <h1>Batches</h1>
...
+			<li>user:u - batches that were created by user u</li>
+			<li>billing_project:bp - batches that were created in billing project bp</li>
```
Comment (danking) [inline batch/batch/front_end/templates/batches.html:29]: change these tabs to spaces and I think we're good to go.

**Review by danking — CHANGES_REQUESTED** [review]

**Review by danking — APPROVED** [review]

**Pull request merged**